extras/tables/table-redis/table_redis.5
.\" Copyright (c) 2015 Emmanuel Vadot <elbarto@bocal.org>
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.\"
.Dd $Mdocdate: May 14 2015 $
.Dt TABLE_REDIS 5
.Os
.Sh NAME
.Nm table_redis
.Nd format description for smtpd redis tables
.Sh DESCRIPTION
This manual page documents the file format of redis tables used by the
.Xr smtpd 8
mail daemon.
.Pp
The format described here applies to tables as defined in
.Xr smtpd.conf 5 .
.Sh REDIS TABLE
A Redis table allows the storing of usernames, passwords, aliases, and domains
in a redis server.
.Pp
The table is used by
.Xr smtpd 8
when authenticating a user, when user information such as user-id and/or
home directory is required for a delivery, when a domain lookup may be required,
and/or when looking for an alias.
.Pp
A Redis table consists of one Redis Databases with one or more keys.
.Pp
If the table is used for authentication, the password should be 
encrypted using the
.Xr crypt 3
function.
Such passwords can be generated using the
.Xr encrypt 1
utility or
.Xr smtpctl 8
encrypt command.
.Sh REDIS TABLE CONFIG FILE
.Cd master
.Dl This is the IP of the master redis server.
.Dl To connect via an unix socket use unix:/path/to/sock
.Dl The default is 127.0.0.1

.Cd master_port
.Dl This is the port used to connect to the master redis server.
.Dl The default is 6379

.Cd slave
.Dl This is the IP of the slave redis server, if any.
.Dl To connect via an unix socket use unix:/path/to/sock

.Cd slave_port
.Dl This is the port used to connect to the slave redis server if any.

.Cd database
.Dl The database number to use.
.Dl The default is 0.

.Cd password
.Dl The password to use to authenticate to the redis server if any.

.Cd query_domain
.Dl This is used to provide a query for a domain query call. All the '%s' are replaced
.Dl with the appropriate data, in this case it would be the right hand side of the SMTP address.
.Dl This expects one string to be returned with a matching domain name.

.Cd query_userinfo
.Dl This is used to provide a query for looking up user information.
.Dl All the '%s' are replaced with the appropriate data, in this case it
.Dl would be the left hand side of the SMTP address.
.Dl This expects three fields to be returned an int containing a UID, an int containing a GID
.Dl and a string containing the home directory for the user.

.Cd query_credentials
.Dl This is used to provide a query for looking up credentials. All the '%s' are replaced
.Dl with the appropriate data, in this case it would be the left hand side of the SMTP address.
.Dl the query expects that there are two strings returned one with a 
.Dl user name one with a password in encrypted format.

.Cd query_alias
.Dl This is used to provide a query to look up aliases. All the '%s' are replaced
.Dl with the appropriate data, in this case it would be the left hand side of the SMTP address.
.Dl This expects one string to be returned with the user name the alias resolves to.
.Dl If the query returns an array, all the data will be concatenated into one string with ',' as a separator
.Pp
.Sh EXAMPLES
Due to the nature of redis, multiple schemas can be used. Those provided here a known to work.
.Pp
.Cd domain
.Dl Since the reuqest must return the domain queried the simple things to do is to have a key
.Dl with the domain name containing the domain name.
.Dl # redis-cli set domains:example.net example.net

.Dl in the redis table configuration file :
.Dl query_domain    GET domains:%s

.Cd userinfo
.Dl Hash works well for users
.Dl # redis-cli HSET user:foo uid 1001
.Dl # redis-cli HSET user:foo gid 1001
.Dl # redis-cli HSET user:foo maildir "/mail/foo"

.Dl in the redis table configuration file :
.Dl query_userinfo  HMGET user:%s uid gid maildir

.Cd credentials
.Dl We can extend the hash for our user to put credential in it
.Dl # redis-cli HSET user:foo login foo
.Dl # redis-cli HSET user:foo passwd encrypted_password

.Dl in the redis table configuration file :
.Dl query_credentials  HMGET user:%s login passwd

.Cd alias
.Dl Using redis sorted list :
.Dl # redis-cli LPUSH aliases:foo@example.net foo
.Dl # redis-cli LPUSH aliases:bar@example.net foo

.Dl in the redis table configuration file :
.Dl query_alias     LRANGE aliases:%s 0 -1

.Sh SEE ALSO
.Xr encrypt 1 ,
.Xr smtpd.conf 5 ,
.Xr smtpctl 8 ,
.Xr smtpd 8


extras/tables/table-redis/table_redis.c
/*	$OpenBSD$	*/

/*
 * Copyright (c) 2013 Eric Faurot <eric@openbsd.org>
 * Copyright (c) 2014 Michael Neumann <mneumann@ntecs.de>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include "includes.h"

#include <sys/types.h>

#include <ctype.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

#include <hiredis.h>

#include "smtpd-defines.h"
#include "smtpd-api.h"
#include "log.h"

enum {
	REDIS_ALIAS = 0,
	REDIS_DOMAIN,
	REDIS_CREDENTIALS,
	REDIS_NETADDR,
	REDIS_USERINFO,
	REDIS_SOURCE,
	REDIS_MAILADDR,
	REDIS_ADDRNAME,

	REDIS_MAX
};

struct config {
	struct dict	 conf;
	redisContext    *master;
	redisContext	*slave;
	char		*queries[REDIS_MAX];
};

static int table_redis_update(void);
static int table_redis_lookup(int, struct dict *, const char *, char *, size_t);
static int table_redis_check(int, struct dict *, const char *);
static int table_redis_fetch(int, struct dict *, char *, size_t);

static redisReply *table_redis_query(const char *key, int service);

static struct config 	*config_load(const char *);
static void		 config_reset(struct config *);
static int		 config_connect(struct config *);
static void		 config_free(struct config *);

static char		*conffile;
static struct config	*config;

int
main(int argc, char **argv)
{
	int	ch;

	log_init(1);
	log_verbose(~0);

	while ((ch = getopt(argc, argv, "")) != -1) {
		switch (ch) {
		default:
			log_warnx("warn: table-redis: bad option");
			return (1);
			/* NOTREACHED */
		}
	}
	argc -= optind;
	argv += optind;

	if (argc != 1) {
		log_warnx("warn: table-redis: bogus argument(s)");
		return (1);
	}

	conffile = argv[0];

	config = config_load(conffile);
	if (config == NULL) {
		log_warnx("warn: table-redis: error parsing config file");
		return (1);
	}
	if (config_connect(config) == 0) {
		log_warnx("warn: table-redis: could not connect");
		return (1);
	}

	table_api_on_update(table_redis_update);
	table_api_on_check(table_redis_check);
	table_api_on_lookup(table_redis_lookup);
	table_api_on_fetch(table_redis_fetch);
	table_api_dispatch();

	return (0);
}

static struct config *
config_load(const char *path)
{
	struct config	*config;
	FILE		*fp;
	size_t		 flen;
	char		*key, *value, *buf, *lbuf;

	lbuf = NULL;

	config = calloc(1, sizeof(*config));
	if (config == NULL) {
		log_warn("warn: table-redis: calloc");
		return (NULL);
	}

	dict_init(&config->conf);

	fp = fopen(path, "r");
	if (fp == NULL) {
		log_warn("warn: table-redis: fopen");
		goto end;
	}

	while ((buf = fgetln(fp, &flen))) {
		if (buf[flen - 1] == '\n')
			buf[flen - 1] = '\0';
		else {
			lbuf = malloc(flen + 1);
			if (lbuf == NULL) {
				log_warn("warn: table-redis: malloc");
				goto end;
			}
			memcpy(lbuf, buf, flen);
			lbuf[flen] = '\0';
			buf = lbuf;
		}

		key = buf;
		while (isspace((unsigned char)*key))
			++key;
		if (*key == '\0' || *key == '#')
			continue;
		value = key;
		strsep(&value, " \t:");
		if (value) {
			while (*value) {
				if (!isspace((unsigned char)*value) &&
				    !(*value == ':' && isspace((unsigned char)*(value + 1))))
					break;
				++value;
			}
			if (*value == '\0')
				value = NULL;
		}

		if (value == NULL) {
			log_warnx("warn: table-redis: missing value for key %s", key);
			goto end;
		}

		if (dict_check(&config->conf, key)) {
			log_warnx("warn: table-redis: duplicate key %s", key);
			goto end;
		}

		value = strdup(value);
		if (value == NULL) {
			log_warn("warn: table-redis: malloc");
			goto end;
		}

		dict_set(&config->conf, key, value);
	}

	free(lbuf);
	fclose(fp);
	return (config);

end:
	free(lbuf);
	if (fp)
		fclose(fp);
	config_free(config);
	return (NULL);
}

static void
config_reset(struct config *config)
{
	size_t	i;

	for (i = 0; i < REDIS_MAX; i++)
		if (config->queries[i]) {
			free(config->queries[i]);
			config->queries[i] = NULL;
		}

	if (config->master) {
		redisFree(config->master);
		config->master = NULL;
	}
}

static int
config_connect(struct config *config)
{
	static const struct {
		const char	*name;
		const char	*default_query;
	} qspec[REDIS_MAX] = {
		{ "query_alias",	"GET alias:%s" },
		{ "query_domain",	"GET domain:%s" },
		{ "query_credentials",	"GET credentials:%s" },
		{ "query_netaddr",	"GET netaddr:%s" },
		{ "query_userinfo",	"GET userinfo:%s" },
		{ "query_source",	"GET source:%s" },
		{ "query_mailaddr",	"GET mailaddr:%s" },
		{ "query_addrname",	"GET addrname:%s" },
	};
	size_t	 i;

	char	*master = "127.0.0.1";
	int	master_port = 6379;
	char	*slave = "NULL";
	int	slave_port = 6380;
	char	*password = NULL;
	int	database;

	char	*q;

	char		*value;
	const char	*e;
	long long	 ll;

	redisReply	*res = NULL;

	log_debug("debug: table-redis: (re)connecting");

	/* Disconnect first, if needed */
	config_reset(config);

	if ((value = dict_get(&config->conf, "master")))
		master = value;
	if ((value = dict_get(&config->conf, "slave")))
		slave = value;

	if ((value = dict_get(&config->conf, "master_port"))) {
		e = NULL;
		ll = strtonum(value, 0, 65535, &e);
		if (e) {
			log_warnx("warn: table-redis: bad value for master_port: %s", e);
			goto end;
		}
		master_port = ll;
	}
	if ((value = dict_get(&config->conf, "slave_port"))) {
		e = NULL;
		ll = strtonum(value, 0, 65535, &e);
		if (e) {
			log_warnx("warn: table-redis: bad value for slave_port: %s", e);
			goto end;
		}
		slave_port = ll;
	}

	if ((value = dict_get(&config->conf, "password")))
	        password = value;

	if ((value = dict_get(&config->conf, "database"))) {
		e = NULL;
		ll = strtonum(value, 0, 256, &e);
		if (e) {
			log_warnx("warn: table-redis: bad value for database: %s", e);
			goto end;
		}
		database = ll;
	}

	if (!strncmp("unix:", master, 5)) {
		log_debug("debug: table-redis: connect via unix socket %s", master + 5);
		config->master = redisConnectUnix(master + 5);
	}
	else {
		log_debug("debug: table-redis: connect to master via tcp at %s:%d", master, master_port);
		config->master = redisConnect(master, master_port);
	}
	if (config->master == NULL) {
		log_warnx("warn: table-redis: Can't create redis context for master");
		goto end;
	}

	if (!config->master->err) {
		log_debug("debug: table-redis: connected to master");
		if (password) {
			res = redisCommand(config->master, "AUTH %s", password);
			if (res->type == REDIS_REPLY_ERROR) {
				log_warnx("warn: table-redis: authentication on master failed");
				goto end;
			}
			freeReplyObject(res);
		}

		if (database != 0) {
			res = redisCommand(config->master, "SELECT %d", database);
			if (res->type != REDIS_REPLY_STATUS) {
				log_warnx("warn: table-redis: database selection on master failed");
				goto end;
			}
			freeReplyObject(res);
		}
	}

	if (slave) {
		if (!strncmp("unix:", slave, 5)) {
			log_debug("debug: table-redis: connect to slave via unix socket %s", slave + 5);
			config->slave = redisConnectUnix(slave + 5);
		}
		else {
			log_debug("debug: table-redis: connect to slave via tcp at %s:%d", slave, slave_port);
			config->slave = redisConnect(slave, slave_port);
		}

		if (config->slave == NULL) {
			log_warnx("warn: table-redis: Can't create redis context for slave");
			goto end;
		}
		if (!config->slave->err) {
			if (password) {
				res = redisCommand(config->slave, "AUTH %s", password);
				if (res->type == REDIS_REPLY_ERROR) {
					log_warnx("warn: table-redis: authentication on slave failed");
					goto end;
				}
				freeReplyObject(res);
			}

			if (database != 0) {
				res = redisCommand(config->slave, "SELECT %d", database);
				if (res->type != REDIS_REPLY_STATUS) {
					log_warnx("warn: table-redis: database selection on slave failed");
					goto end;
				}
				freeReplyObject(res);
			}
		}
	}

	for (i = 0; i < REDIS_MAX; i++) {
		q = dict_get(&config->conf, qspec[i].name);
		if (q)
			config->queries[i] = strdup(q);
		else
			config->queries[i] = strdup(qspec[i].default_query);
		if (config->queries[i] == NULL) {
			log_warn("warn: table-redis: malloc");
			goto end;
		}
	}

	if (config->master->err && config->slave->err) {
		log_warnx("warn: table-redis: redisConnect for master and slave failed");
		goto end;
	}

	log_debug("debug: table-redis: connected");

	return (1);

end:
	if (res)
		freeReplyObject(res);
	config_reset(config);
	return (0);
}

static void
config_free(struct config *config)
{
	void	*value;

	config_reset(config);

	while (dict_poproot(&config->conf, &value))
		free(value);

	free(config);
}

static int
table_redis_update(void)
{
	struct config	*c;

	if ((c = config_load(conffile)) == NULL)
		return (0);
	if (config_connect(c) == 0) {
		config_free(c);
		return (0);
	}

	config_free(config);
	config = c;
	return (1);
}

static redisReply *
table_redis_query(const char *key, int service)
{
	redisReply	*res;
	char		*query = NULL;
	int		i;
	int		retry_times;

	retry_times = 3;

 retry:
	--retry_times;
	if (retry_times < 0) {
		log_warnx("warn: table-redis: giving up: too many retries");
		return (NULL);
	}

	for(i = 0; i < REDIS_MAX; i++)
		if (service == 1 << i) {
			query = config->queries[i];
			break;
		}

	if (query == NULL)
		return (NULL);

	if (!config->master->err) {
		log_debug("debug: table-redis: running query \"%s\" on master", query);
		res = redisCommand(config->master, query, key);
	}
	else if (!config->slave->err) {
		log_debug("debug: table-redis: running query \"%s\" on slave", query);
		res = redisCommand(config->slave, query, key);
	}
	else
		return (NULL);
	if (res == NULL) {
		log_warnx("warn: table-redis: redisCommand: %s",
		    config->master->errstr);

		if (config_connect(config))
			goto retry;

		return (NULL);
	}

	return (res);
}

static int
table_redis_check(int service, struct dict *params, const char *key)
{
	int		 r;
	redisReply	*reply;

	if (config->master == NULL && config_connect(config) == 0)
		return (-1);

	reply = table_redis_query(key, service);
	if (reply == NULL)
		return (-1);

	switch (reply->type) {
		case REDIS_REPLY_INTEGER:
		case REDIS_REPLY_STRING:
		case REDIS_REPLY_ARRAY:
			r = 1;
			break;

		case REDIS_REPLY_NIL:
			r = 0;
			break;

		case REDIS_REPLY_STATUS:
		case REDIS_REPLY_ERROR:
		default:
			r = -1;
			break;
	}

	freeReplyObject(reply);

	return (r);
}

static int
table_redis_lookup(int service, struct dict *params, const char *key, char *dst, size_t sz)
{
	redisReply	*reply, *elmt;
	unsigned int	i;
	int		r;

	if (config->master == NULL && config_connect(config) == 0)
		return (-1);

	reply = table_redis_query(key, service);
	if (reply == NULL)
		return (-1);

	r = 1;
	switch(service) {
	case K_ALIAS:
		memset(dst, 0, sz);
		if (reply->type == REDIS_REPLY_STRING) {
			if (strlcat(dst, reply->str, sz) >= sz) {
				log_warnx("warn: table-redis: result too large");
				r = -1;
			}
		}
		else if (reply->type == REDIS_REPLY_ARRAY) {
			if (reply->elements == 0)
				r = 0;

			for (i = 0; i < reply->elements; i++) {
				elmt = reply->element[i];
				if (elmt == NULL ||
				    elmt->type != REDIS_REPLY_STRING) {
					r = -1;
					break;
				}
				if (dst[0] && strlcat(dst, ", ", sz) >= sz) {
					log_warnx("warn: table-redis: result too large");
					r = -1;
				}
				if (strlcat(dst, elmt->str, sz) >= sz) {
					log_warnx("warn: table-redis: result too large");
					r = -1;
				}
			}
		}
		else
			r = -1;
		break;
	case K_CREDENTIALS:
	case K_USERINFO:
		memset(dst, 0, sz);
		if (reply->type == REDIS_REPLY_STRING) {
			if (strlcpy(dst, reply->str, sz) >= sz) {
				log_warnx("warn: table-redis: result too large");
				r = -1;
			}
		}
		else if (reply->type == REDIS_REPLY_ARRAY) {
			if (reply->elements == 0)
				r = 0;

			for (i = 0; i < reply->elements; i++) {
				elmt = reply->element[i];
				if (elmt == NULL ||
				    elmt->type != REDIS_REPLY_STRING) {
					r = -1;
					break;
				}
				if (dst[0] && strlcat(dst, ":", sz) >= sz) {
					log_warnx("warn: table-redis: result too large");
					r = -1;
				}
				if (strlcat(dst, elmt->str, sz) >= sz) {
					log_warnx("warn: table-redis: result too large");
					r = -1;
				}
			}
		}
		else
			r = -1;
		break;
	case K_DOMAIN:
	case K_NETADDR:
	case K_SOURCE:
	case K_MAILADDR:
	case K_ADDRNAME:
		if (reply->type == REDIS_REPLY_STRING) {
			if (strlcpy(dst, reply->str, sz) >= sz) {
				log_warnx("warn: table-redis: result too large");
				r = -1;
			}
		}
		else
			r = -1;
		break;
	default:
		log_warnx("warn: table-redis: unknown service %d",
		    service);
		r = -1;
	}

	freeReplyObject(reply);
	return (r);
}

static int
table_redis_fetch(int service, struct dict *params, char *dst, size_t sz)
{
	return (-1);
}
